app_flutter/lib/task_matrix.dart

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+
+import 'package:cocoon_service/protos.dart' show CommitStatus, Stage, Task;
+
+/// Class to handle data operations on [List<CommitStatus>].
+///
+/// Flattens the mapping of one [CommitStatus] from many [Stage] objects,
+/// where each [Stage] object maps to many [Task] objects, to a 2D matrix.
+/// Simplifies logic further by moving related [Task] into a [Column], so the
+/// matrix is composed of just a list of columns.
+///
+/// After construction, the underlying matrix and statuses do not get modified.
+/// Only the column mapping is changed with additional operations.
+class TaskMatrix {
```

Comment:
Spoke offline about this - the details and documentation are very solid, but I would like a high-level overview of this algorithm to make it easier for people unfamiliar with it to get up to speed quickly

Reply:
I updated the overview and expanded the discussion on the sorting algorithm. Let me know if there's other parts that should be covered

Reply:
Updated again using Jonah's reference of [the Wikipedia Path-based strong component algorithm page](https://en.wikipedia.org/wiki/Path-based_strong_component_algorithm).